CASSANDRA-18345 Stream all components registered by an SSTable

pkolaczk opened the pull request:
Custom indexes may register custom components.
We need to include them in streaming as well.

Commit:
Stream all components registered by an sstable

Indexes may register custom components.
We need to include them in streaming to avoid costly
building of indexes on the receiving side.

In order to make this work,
zero copy streaming had to be modified to index its writers
by component name instead of component type because
component types are not unique - e.g. many index components
are of type CUSTOM.

maedhroz commented:
nit: There are some unused import errors on J11 checkstyle

maedhroz commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,134 @@
+{
+    private static final ByteBuffer BLOB = ByteBuffer.wrap(new byte[1 << 16]);
+    // zero copy streaming sends all components, so the events will include non-Data files as well
+    private static final int NUM_COMPONENTS = 17;
```
nit: We can probably make this a little more self-documenting if we sum up `V1OnDiskFormat.PER_SSTABLE_COMPONENTS`, `V1OnDiskFormat.LITERAL_COMPONENTS`, and the expected core SSTable components (perhaps from `SSTableFormat#streamingComponents()` sans compression info, which I guess we don't have here?)

pkolaczk replied:
Good idea. Fixed

maedhroz commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,134 @@
+                "CREATE TABLE %s.test (pk int PRIMARY KEY, v text, b blob) WITH compression = { 'enabled' : false };"
+            ));
+            cluster.schemaChange(withKeyspace(
+                "CREATE CUSTOM INDEX ON %s.test(v) USING 'StorageAttachedIndex';"
```
Once numerics support merges, I guess we'll want to parameterize this test further to use those. The number of expected index components might not actually change though. CC @mike-tr-adamson 

I'm not sure if that means we should merge this first and adjust inside a rebased CASSANDRA-18067 or the other way around.

maedhroz commented on src/java/org/apache/cassandra/io/sstable/SSTableZeroCopyWriter.java:
```diff
@@ -61,12 +58,8 @@ public SSTableZeroCopyWriter(Builder<?, ?> builder,
-        if (!descriptor.getFormat().streamingComponents().containsAll(components))
-            throw new AssertionError(format("Unsupported streaming component detected %s",
-                                            Sets.difference(ImmutableSet.copyOf(components), descriptor.getFormat().streamingComponents())));
-
```
@pkolaczk @mike-tr-adamson I'm not sure we want to just remove this validation, but checking SAI components raises some other questions. The simplest way around this for now might be...

```
Set<Component> unsupported = components.stream()
                                       .filter(c -> !descriptor.getFormat().streamingComponents().contains(c))
                                       .filter(c-> c.type != SSTableFormat.Components.Types.CUSTOM)
                                       .collect(Collectors.toSet());

if (!unsupported.isEmpty())
    throw new AssertionError(format("Unsupported streaming component detected %s", unsupported));
```

However, it might also be worth looking at why we use `CUSTOM` for SAI components in the first place. Is now the time to simply create a new first-class type in `Types`?

```
public static final Component.Type SAI = Component.Type.create("SAI", "SAI\\+.*.db", null);
```

(The regex could be stricter, but details...)

If we do this, we the `CUSTOM` in the check above becomes `SAI`, we replace `CUSTOM` w/ `SAI` in a few places in the SAI code, like `Version`, and things should just work, right?

maedhroz replied:
btw, I just did exactly this, and it works fine...at least `IndexStreamingTest` has no issues

pkolaczk replied:
Why do we want to artificially restrict the set of components streamed in the place that can stream any component?
It's not like streaming supports only some kind of components and doesn't support some other types. It will copy anything we'll throw at it.

pkolaczk replied:
As for this:
```
public static final Component.Type SAI = Component.Type.create("SAI", "SAI\\+.*.db", null);
```

This wouldn't be very consistent with what we use component types for now in the core - because now each core sstable component  has its own type. So why should we throw all different SAI component types into a single type? 

I think we need to take another look at how custom components are registered and simply allow modules to provide their own types. So SAI component types should be defined in SAI. We could also allow the component type to decide whether it should be streamed or not.

maedhroz replied:
> I think we need to take another look at how custom components are registered and simply allow modules to provide their own types. So SAI component types should be defined in SAI. We could also allow the component type to decide whether it should be streamed or not.

I think this would work. +1

pkolaczk replied:
Fixed

maedhroz commented on src/java/org/apache/cassandra/db/streaming/ComponentManifest.java:
```diff
@@ -52,13 +53,17 @@ public ComponentManifest(Map<Component, Long> components)
+        for (Component component : streamingComponents)
         {
-            File file = descriptor.fileFor(component);
+            if (component == SSTableFormat.Components.TOC)
```
nit: I guess the spirit of the original code here was to delegate this as much as possible to the SSTable format. What if we tried to preserve that (and coupled it with the `SAI` component I suggested below) and did something like...

```
if (!sstable.descriptor.getFormat().streamingComponents().contains(component) && component.type != BigFormat.Components.Types.SAI)
```
I like the idea of using `sstable.getComponents()` to make sure we don't forget anything, but it seems like picking out the `TOC` alone is a little brittle. WDYT?

maedhroz replied:
We could also hide this in `SSTable` with a `getStreamingComponents()` and leave this logic in the manifest almost unchanged? It already makes a copy of `components`, and this isn't really a hot path anyway, so filtering there w/ something like...

```
public Set<Component> getStreamingComponents()
{
    return components.stream()
                     .filter(c -> descriptor.getFormat().streamingComponents().contains(c) || c.type == Components.Types.SAI)
                     .collect(Collectors.toSet());
}
```

pkolaczk replied:
First, I like the idea of having `getStreamingComponents` method on the sstable. This looks like a better place to tell which components are eligible for streaming. 

However, even with such method, we still have a "responsibility leak" from SAI here. 

```
c.type == Components.Types.SAI
```

Why should SSTable code know anything about SAI? Why should it assume all SAI components should be streamed?  IMHO this is something that should be decided by the SAI code (or whoever owns the component), not by sstable code.

SAI already controls the type of the component, so I guess the ideal-world solution would be to have this information associated directly with the type so instead of special-casing SAI vs core components, we could write the condition like this:

```
 return components.stream()
                     .filter(c -> c.type.shouldBeStreamed)
                     .collect(Collectors.toSet());
```

WDYT?

maedhroz replied:
Yeah, at the end of the day, some things are just not part of the SSTable format, so if/when that is the case, having the component decide whether it's "streamable" doesn't seem like a bad idea. (Or we could do that for all components.)

maedhroz reviewed (changes requested):
The approach looks fine here. There are just a few things we should at least double check in terms of how we generate and validate the components we want to stream (...and we have to decide whether to merge this now and make the numerics patch add some testing or wait for numerics and test in this patch).

Commit:
Address non-controversial review comments

Commit:
Readd validation of components in SSTableZeroCopyWriter

pkolaczk commented:
I addressed all review comments, except the part about introducing the SAI type for SAI components, because IMHO that needs a wider discussion.

Commit:
Allow component types to decide whether they need to be streamed

Instead of hardcoding a collection of streamable components,
a flag `streamable` is added to the component type.

This allows custom components (e.g. SAI components) to decide whether
they should be streamed.

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -59,7 +60,18 @@ public interface SSTableFormat<R extends SSTableReader, W extends SSTableWriter>
+     * Custom components registered by e.g. secondary indexes are not included.
+     * Use {@link SSTableReader#getStreamingComponents()} for the list of all components including the custom ones.
+     */
+    default Set<Component> streamingComponents()
```
Once all `Component` instances have their "streamability" embedded, we don't need this method at all, right? i.e. `SSTableZeroCopyWriter` will no longer need special casing?

pkolaczk replied:
Looks like we need it for testing. I'll think of a way to get rid of it completely.

pkolaczk replied:
I simply moved to testing code.

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -156,23 +168,23 @@ public static class Types
+            public static final Component.Type TOC = Component.Type.createSingleton("TOC", "TOC.txt", false, null);
             // built-in secondary index (may exist multiple per sstable)
-            public static final Component.Type SECONDARY_INDEX = Component.Type.create("SECONDARY_INDEX", "SI_.*.db", null);
+            public static final Component.Type SECONDARY_INDEX = Component.Type.create("SECONDARY_INDEX", "SI_.*.db", true, null);
```
nit: Are `SECONDARY_INDEX` components streamable? I thought we always rebuilt them at the receiver...

pkolaczk replied:
Good catch. I forgot to change it.

pkolaczk replied:
Fixed

maedhroz commented on src/java/org/apache/cassandra/io/sstable/format/SSTableFormat.java:
```diff
@@ -156,23 +168,23 @@ public static class Types
+            public static final Component.Type SECONDARY_INDEX = Component.Type.create("SECONDARY_INDEX", "SI_.*.db", true, null);
             // custom component, used by e.g. custom compaction strategy
-            public static final Component.Type CUSTOM = Component.Type.create("CUSTOM", null, null);
+            public static final Component.Type CUSTOM = Component.Type.create("CUSTOM", null, true, null);
```
I guess not all CUSTOM components should be streamable?

Is the plan here to just create something like the other `Types` inner classes, but in SAI's `V1OnDiskFormat`? If we did that, we could specify "streamability" there in those types. For example:

`Component.Type TERMS_DATA = Component.Type.create("TermsData", "SAI\\+.*\\+TermsData.db", true, null);`

This is really where all the stuff in `IndexComponent` probably should have gone in the first place.

CC @mike-tr-adamson

maedhroz replied:
...or more completely, something like...

```
public static class Components
{
    public static class Types extends AbstractSSTableFormat.Components.Types
    {
        public static final Component.Type TERMS_DATA = Component.Type.create("TermsData", "SAI\\+.*\\+TermsData.db", true, null);
...
    }

    public final static Component TERMS_DATA = Types.TERMS_DATA.getSingleton();
...
}
```

pkolaczk replied:
Yes, that is the plan. I haven't finished that yet.

pkolaczk replied:
Fixed

Commit:
Remove SSTableFormat#streamingComponents

We don't need it because we can check streamability by inspecting
component's streamable field directly.

adelapena commented on src/java/org/apache/cassandra/io/sstable/format/big/BigFormat.java:
```diff
@@ -82,9 +82,9 @@ public static class Components extends SSTableFormat.Components
         {
             // index of the row keys with pointers to their positions in the data file
-            public static final Component.Type PRIMARY_INDEX = Component.Type.createSingleton("PRIMARY_INDEX", "Index.db", BigFormat.class);
+            public static final Component.Type PRIMARY_INDEX = Component.Type.createSingleton("PRIMARY_INDEX", "Index.db", true, BigFormat.class);
```
We'll need to update [the doc on `SSTable_API.md`](https://github.com/apache/cassandra/blob/cep-7-sai/src/java/org/apache/cassandra/io/sstable/SSTable_API.md#components) to include the new `streamable` parameter.

pkolaczk replied:
Fixed

adelapena commented on src/java/org/apache/cassandra/io/sstable/Component.java:
```diff
@@ -60,31 +61,34 @@ public final static class Type
+        public static Type create(String name, String repr, boolean isStreamable, Class<? extends SSTableFormat<?, ?>> formatClass)
         {
-            return new Type(name, repr, false, formatClass);
+            return new Type(name, repr, false, isStreamable, formatClass);
```
Nit: the parameter could be named `streamable` for consistency with the attribute and constructor names.

pkolaczk replied:
Fixed

adelapena commented on src/java/org/apache/cassandra/io/sstable/Component.java:
```diff
@@ -60,31 +61,34 @@ public final static class Type
+         * @param formatClass  format class for which this type is defined for
          */
-        public static Type createSingleton(String name, String repr, Class<? extends SSTableFormat<?, ?>> formatClass)
+        public static Type createSingleton(String name, String repr, boolean isStreamable, Class<? extends SSTableFormat<?, ?>> formatClass)
```
Nit: the parameter could be named `streamable` for consistency with the attribute and constructor names.

pkolaczk replied:
Fixed

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,151 @@
+    {
+        try (Cluster cluster = init(Cluster.build(2)
+                                           .withConfig(c -> c.with(Feature.values())
+                                                             .set("stream_entire_sstables", zeroCopyStreaming).set("streaming_slow_events_log_timeout", "0s"))
```
Nit: break the long line
```suggestion
                                                             .set("stream_entire_sstables", zeroCopyStreaming)
                                                             .set("streaming_slow_events_log_timeout", "0s"))
```

pkolaczk replied:
Indeed. This test has been mostly copied from RebuildStreamingTest. Those long lines are there as well.
Do you want me to cleanup long lines in there as well?

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,151 @@
+                                                             .set("stream_entire_sstables", zeroCopyStreaming).set("streaming_slow_events_log_timeout", "0s"))
+                                           .start()))
+        {
+            // streaming sends events every 65k, so need to make sure that the files are larger than this to hit
+            // all cases of the vtable
```
Maybe this comment would be better placed either right before the `INSERT` or when declaring the `BLOB` size.

pkolaczk replied:
Yup, moving it makes sense. Moved to above BLOB.

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,151 @@
+            assertThat(qr.hasNext()).isTrue();
+            Row row = qr.next();
+            QueryResultUtil.assertThat(row)
+                           .isEqualTo("peers", Collections.singletonList("/127.0.0.2:7012"))
```
```suggestion
                           .isEqualTo("peers", Collections.singletonList(second.broadcastAddress().toString()))
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,151 @@
+            assertThat(qr.hasNext()).isTrue();
+
+            QueryResultUtil.assertThat(qr.next())
+                           .isEqualTo("peers", Collections.singletonList("/127.0.0.1:7012"))
```
```suggestion
                           .isEqualTo("peers", Collections.singletonList(first.broadcastAddress().toString()))
```

adelapena commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,151 @@
+                           .columnsEqualTo("files_sent", "files_to_send", "bytes_sent", "bytes_to_send").isEqualTo("files_sent", 0L);
+
+            // did we trigger slow event log?
+            cluster.forEach(i -> Assertions.assertThat(i.logs().grep("Handling streaming events took longer than").getResult()).describedAs("Unable to find slow log for node%d", i.config().num()).isNotEmpty());
```
Nit: break the long line
```suggestion
            cluster.forEach(i -> Assertions.assertThat(i.logs().grep("Handling streaming events took longer than").getResult())
                                           .describedAs("Unable to find slow log for node%d", i.config().num())
                                           .isNotEmpty());
```

adelapena commented on src/java/org/apache/cassandra/io/sstable/SSTableZeroCopyWriter.java:
```diff
@@ -195,14 +197,15 @@ public void close()
+    public void writeComponent(Component component, DataInputPlus in, long size) throws ClosedChannelException
     {
-        logger.info("Writing component {} to {} length {}", type, componentWriters.get(type).getPath(), prettyPrintMemory(size));
+        SequentialWriter writer = componentWriters.get(component.name);
```
Build fails here with `eclipse-warnings` saying `Potential resource leak: 'writer' may not be closed`

pkolaczk replied:
That looks like a false positive to me. It isn't supposed to be closed here. All writers are closed in the `close()` method. How do I suppress this?

adelapena replied:
You can add `@SuppressWarnings("resource")` with a brief comment about where it will be closed. We have a bunch of those, for example [this one](https://github.com/apache/cassandra/blob/eda0bcbefaf38e4069ac6db3a00937d778090083/src/java/org/apache/cassandra/service/reads/range/RangeCommands.java#L55).

pkolaczk replied:
I'm not getting this error when building. And the project has built fine in CircleCI. 
I tried adding `@SuppressWarnings("resource")` but intellij immediately marks it as redundant :shrug:

adelapena replied:
I get this error locally when running `eclipse-warnings` with JDK 8:
```
eclipse-warnings:
    [mkdir] Created dir: /Users/adelapena/src/cassandra/trunk/build/ecj
     [echo] Running Eclipse Code Analysis.  Output logged to /Users/adelapena/src/cassandra/trunk/build/ecj/eclipse_compiler_checks.txt
     [java] ----------
     [java] 1. ERROR in /Users/adelapena/src/cassandra/trunk/src/java/org/apache/cassandra/io/sstable/SSTableZeroCopyWriter.java (at line 202)
     [java] 	SequentialWriter writer = componentWriters.get(component.name);
     [java] 	                 ^^^^^^
     [java] Potential resource leak: 'writer' may not be closed
     [java] ----------
     [java] 1 problem (1 error)

BUILD FAILED
```
CircleCI also hits it for j8: https://app.circleci.com/pipelines/github/adelapena/cassandra/2986/workflows/82e3ccc4-83ce-49c8-ab0d-036d100cae8e/jobs/54244

However, it doesn't fail for j11 builds: https://app.circleci.com/pipelines/github/adelapena/cassandra/2986/workflows/554d79db-c157-4e53-944a-ced1dffc242f

Do you have a link to the CircleCI run? Maybe it was a j11-only run? j8 support will we dropped very soon ([CASSANDRA-18255](https://issues.apache.org/jira/browse/CASSANDRA-18255)), but we still need it.

pkolaczk replied:
Ah! Gotcha. So that's the difference: JDK 8. I was using 11. Checking with JDK 8....

maedhroz replied:
If you want IDEA to stop complaining, use `@SuppressWarnings({ "resource", "RedundantSuppression" })`

pkolaczk replied:
fixed

Commit:
Use distinct component types for SAI components

Don't use CUSTOM component.
Instead, create separate component type for each SAI index component type.

maedhroz commented on src/java/org/apache/cassandra/io/sstable/SSTableZeroCopyWriter.java:
```diff
@@ -61,12 +61,14 @@ public SSTableZeroCopyWriter(Builder<?, ?> builder,
+                                               .filter(c -> !c.type.streamable)
+                                               .collect(Collectors.toSet());
+        if (!unsupported.isEmpty())
+            throw new AssertionError(format("Unsupported streaming component detected %s", unsupported));
```
```suggestion
            throw new AssertionError(format("Unsupported streaming components detected: %s", unsupported));
```

maedhroz commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,151 @@
+                         + V1OnDiskFormat.LITERAL_COMPONENTS.size();
+    }
+
+    private static int sstableStreamingComponentsCount() {
```
nit: newline for `{`

pkolaczk replied:
Fixed

maedhroz commented on test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java:
```diff
@@ -0,0 +1,151 @@
+        DatabaseDescriptor.clientInitialization();
+        NUM_COMPONENTS = sstableStreamingComponentsCount()
+                         + V1OnDiskFormat.PER_SSTABLE_COMPONENTS.size()
+                         + V1OnDiskFormat.LITERAL_COMPONENTS.size();
```
Nice. We'll expand on this whenever numerics merge.

CC @mike-tr-adamson

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/format/IndexComponent.java:
```diff
@@ -79,10 +84,20 @@ public enum IndexComponent
+    private static Component.Type componentType(String name)
+    {
+        return Component.Type.create(SAI_DESCRIPTOR + SAI_SEPARATOR + name.toUpperCase(),
+                                     SAI_DESCRIPTOR + ".*." + name + ".db",
```
Looking at `Version.defaultFileNameFormat()`, I think you'd want something like:

`SAI_DESCRIPTOR + "\\+.*\\+" + name + ".db"`

`".*."` doesn't need the second period?

pkolaczk replied:
Fixed, and also added pattern quoting.

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/format/IndexComponent.java:
```diff
@@ -79,10 +84,20 @@ public enum IndexComponent
+
+    private static Component.Type componentType(String name)
+    {
+        return Component.Type.create(SAI_DESCRIPTOR + SAI_SEPARATOR + name.toUpperCase(),
```
nit: `toUpperCase()` here seems unnecessary, as it will give us weird things like "GROUPCOMPLETE". Camel is ok?

pkolaczk replied:
Indeed. Fixed

maedhroz commented on src/java/org/apache/cassandra/index/sai/disk/format/IndexComponent.java:
```diff
@@ -79,10 +84,20 @@ public enum IndexComponent
      */
     GROUP_COMPLETION_MARKER("GroupComplete");
 
+    public final Component.Type type;
```
super OCD nit: maybe `type` could go below `name` so they're in the same order in the ctor :D

pkolaczk replied:
+1

maedhroz reviewed (approved):
Made a few more nits, but this looks pretty nice now.

Again, we'll have to modify/add some things here to deal w/ the numerics patch when it's ready (or in that patch if this merges), but that shouldn't be hard.

Commit:
Add streamable param to code samples in SSTable_API.md

Commit:
Rename isStreamable -> streamable

pkolaczk committed:
Split long lines

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

Commit:
Move comment to a better place

pkolaczk committed:
Update test/distributed/org/apache/cassandra/distributed/test/sai/IndexStreamingTest.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Use broadcastAddress instead of hardcoded IP

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Split

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Fix assertion message

Co-authored-by: Caleb Rackliffe <maedhroz@users.noreply.github.com>

Commit:
Fix formatting

Commit:
Improve SAI component regex

Commit:
Suppress eclipse-warnings about resource not being closed

pkolaczk commented:
CI run: https://app.circleci.com/pipelines/github/pkolaczk/cassandra/16/workflows/41cc8da8-cf2b-4985-be0c-cf5ffd9598d8

maedhroz commented on src/java/org/apache/cassandra/io/sstable/SSTableZeroCopyWriter.java:
```diff
@@ -195,14 +197,16 @@ public void close()
+    public void writeComponent(Component component, DataInputPlus in, long size) throws ClosedChannelException
     {
-        logger.info("Writing component {} to {} length {}", type, componentWriters.get(type).getPath(), prettyPrintMemory(size));
+        @SuppressWarnings({"resource", "RedundantSuppression"})  // all writers are closed in close()
```
Curious...can individual writers fail and cause some writers not to be closed? (above, on 197)

adelapena replied:
I guess if that's the case, we could use `FBUtitilites#closeAll` at `SSTableZeroCopyWriter#close`. But we are not modifying that here, so I think it can be done on a separate ticket.

maedhroz commented:
Committed as https://github.com/apache/cassandra/commit/5d3f257477cba2d7f33f842dba4582d0660f5738
